File: src/javascript/algorithms/strings/greatestCommonDivisor.js
// 1. BrutForce (Time complexity: O(min(m,n) . (m+n)), Space complexity: O(min(m,n)))
function gcdOfStrings(string1, string2) {
    let l1 = string1.length;
    let l2 = string2.length;

    for(let i= Math.min(l1,l2); i >= 1; i--) {
        if(isValidGcdString(string1, string2, i)) {
            return string1.substr(0, i);
        }
    }
}

function isValidGcdString(string1, string2, k) {
    let l1 = string1.length;
    let l2 = string2.length;

    if(l1%k > 0 || l2%k >0) {
        return false;
    }
    let base = string1.substr(0, k);
    return string1.replaceAll(base, "").length === 0 && string2.replaceAll(base, "").length === 0;
}

console.log(gcdOfStrings("AB", "AB"));
console.log(gcdOfStrings("ABCABCABC", "ABCABC"));
console.log(gcdOfStrings("ABABAB", "ABAB"));

//2. GCD using Euclidean's algorithm
function gcdOfStrings1(string1, string2){
    //Verify whether GCD strings exists or not
    if(!(string1+string2 === string2+string1)) {
        return false;
    }

    let gcdLength = gcd(string1.length, string2.length);
    return string1.substr(0, gcdLength);
}

function gcd(x, y) {
    if(y === 0) {
        return x;
    }
    return gcd(y, x%y);
}

console.log(gcdOfStrings1("AB", "AB"));
console.log(gcdOfStrings1("ABCABCABC", "ABCABC"));
console.log(gcdOfStrings1("ABABAB", "ABAB"));

File: src/java1/algorithms/strings/greatestCommonDivisor/GreatestCommonDivisor.java
package java1.algorithms.strings.greatestCommonDivisor;

public class GreatestCommonDivisor {
    //1. GCD using Euclidean's algorithm TC: O(n+m)(i.e, O(n + m + log(min(n, m)))), SC: O(n+m)
    private static String gcdOfStrings1(String str1, String str2) {
        if(!((str1+str2).equals(str2+str1))) return "";

        int gcdLength = gcd(str1.length(), str2.length());

        return str1.substring(0, gcdLength);

    }

    private static int gcd(int a, int b){
        return b == 0 ? a : gcd(b, a%b);
    }

    // 2. BrutForce (Time complexity: O(min(m,n) . (m+n)), Space complexity: O(min(m,n)))
    private static String gcdOfStrings2(String str1, String str2) {
        int l1 = str1.length(), l2 = str2.length();

        for(int i = Math.min(l1, l2); i >=1; i++){
            if(hasGcdString(str1, str2, i)) {
                return str1.substring(0, i);
            }
        }

        return "";
    }

    private static boolean hasGcdString(String str1, String str2, int k){
        int l1 = str1.length(), l2 = str2.length();

        if(l1%k > 0 || l2%k > 0){
            return false;
        }

        int f1 = l1/k, f2 = l2/k;
        String base = str1.substring(0, k);

        return base.repeat(f1) == str1 && base.repeat(f2) == str2;
    }
    
    public static void main(String[] args) {
        System.out.println(gcdOfStrings1("AB", "AB"));
        System.out.println(gcdOfStrings1("ABCABCABC", "ABCABC"));
        System.out.println(gcdOfStrings1("ABABAB", "AB"));

        System.out.println(gcdOfStrings2("AB", "AB"));
        System.out.println(gcdOfStrings2("ABCABCABC", "ABCABC"));
        System.out.println(gcdOfStrings2("ABABAB", "AB"));
    }
}

